### PR TITLE
YM-16657: Update Exposed Errors to allow more descriptive feedback in Objc implementations.

### PR DESCRIPTION
**Pull request by mkotsiandris**
## Description

Currently, the Error types exposed by the SDK don't have the capability to provide a detailed description of the issue. Especially for integrations on Objc, the bridge between Swift and Objc resulted in error codes limited to the integer assigned to the respective enum case. 

This PR aims to mitigate this issue by restructuring and exposing more descriptive Error types. 

## Approach
We exposed three different types of Errors `SetupError`, `ShareRequestError`, `CallbackBackendError` based on the three different sections this SDK is going to perform on.

As a result, we have:
- `SetupError` provides feedback for misconfiguration of the project.
- `ShareRequestError` provides feedback for potential issues during the share request
- `CallbackBackendError` provides feedback after the completion of the share request and during the execution of the callback to the backend.

### CODE REVIEW

File: YotiButtonSDK/YotiSDK.swift [hunk 1]
```diff
@@ -57,11 +53,11 @@ public class YotiSDK: NSObject {
     func startScenario(for useCaseID: String, with delegate: SDKDelegate) throws {
 
         guard let scenario = scenario(for: useCaseID) else {
-            throw GenericError.nilValue("scenario")
+            throw ShareRequestError.startScenarioError("No ScenarioID found")
```
Comment (johnmillz) [inline YotiButtonSDK/YotiSDK.swift]:
```suggestion
            throw ShareRequestError.startScenarioError("No ScenarioID associated with this useCaseID")
```

File: Sample/Cartfile.resolved [hunk 2]
```diff
@@ -1 +1 @@
-github "getyoti/ios-sdk-button" "2.0"
+github "getyoti/ios-sdk-button" "3.0"
```
Comment (johnmillz) [inline Sample/Cartfile.resolved]:
```suggestion
github "getyoti/ios-sdk-button" "3.0.0"
```

or depending on the release tag.

```suggestion
github "getyoti/ios-sdk-button" "v3.0.0"
```

File: Sample/Cartfile [hunk 3]
```diff
@@ -1 +1 @@
-github "getyoti/ios-sdk-button" ~> 2.0
+github "getyoti/ios-sdk-button" ~> 3.0
```
Comment (johnmillz) [inline Sample/Cartfile]:
```suggestion
github "getyoti/ios-sdk-button" ~> 3.0.0
```

File: YotiButtonSDK/Kernel/KernelSDK.swift [hunk 4]
```diff
@@ -55,21 +49,21 @@ class KernelSDK: NSObject {
...
                 return
             }
 
             guard UIApplication.shared.canOpenURL(url) else {
-                delegate.yotiSDKDidFail(for: useCaseID, with: GenericError.unknown(url.absoluteString))
+                delegate.yotiSDKDidFail(for: useCaseID, with: SetupError.invalidApplicationQueriesSchemes(url))
 
                 print("Cannot Open Yoti, please check LSApplicationQueriesSchemes or install Yoti")
                 return
             }
```
Comment (johnmillz) [inline YotiButtonSDK/Kernel/KernelSDK.swift]: Not sure this check is quite valid as it stands. This currently only checks we can open a URL which in practice doesn't ever use the `yoti://` scheme referenced in the print statement and thus error. To check the `LSApplicationQueriesSchemes` we can read it from the bundle as above with the URL schemes: `Bundle.main.object(forInfoDictionaryKey: "LSApplicationQueriesSchemes")`.  

I think the intention was to use this to check the ID Application (Yoti) is installed so we could split this:

A check prior to making the request to check the `LSApplicationQueriesSchemes` as part of the setup error.
A check on making the request with a dummy URL including the sceme to probe for the ID Applications.

Not a fan of the dummy URL but it satisfies what this apparently was attempting to do, I think - I could equally be completly wrong...

**Review by johnmillz — APPROVED** [review]

**Review by CharlesVu — APPROVED** [review]

**Pull request merged**